### PR TITLE
Support for Android NDK r22

### LINKED ISSUE #13685: Fail to configure with Android NDK r22
No configdata.pm generated.

OpenSSL Version: 1.1.1i

Analysis:

The condition on https://github.com/openssl/openssl/blob/OpenSSL_1_1_1i/Configurations/15-android.conf#L33 doesn't meet anymore. The new NDK doesn't have `platforms` folder and `AndroidVersion.txt`.

```
$ ./Configure android-x86 no-shared                                                                                                                                                                      
Configuring OpenSSL version 1.1.1i (0x1010109fL) for android-x86
Using os-specific seed configuration
$ANDROID_NDK_HOME=/opt/android-ndk is invalid at (eval 10) line 37.
```

```
$ export
ANDROID_NDK=/opt/android-ndk
ANDROID_NDK_HOME=/opt/android-ndk
COLORFGBG='15;0'
COLORTERM=truecolor
DBUS_SESSION_BUS_ADDRESS='unix:path=/run/user/1000/bus'
DESKTOP_SESSION=/usr/share/xsessions/plasma
DISPLAY=:0
EDITOR=vim
GIT_ASKPASS=/usr/bin/ksshaskpass
GOPATH=/home/wong/go
GRADLE_HOME=/usr/share/java/gradle
GTK2_RC_FILES=/etc/gtk-2.0/gtkrc:/home/wong/.gtkrc-2.0:/home/wong/.config/gtkrc-2.0
GTK_IM_MODULE=fcitx
GTK_MODULES=canberra-gtk-module
GTK_RC_FILES=/etc/gtk/gtkrc:/home/wong/.gtkrc:/home/wong/.config/gtkrc
HISTFILE=/home/wong/.zhistory
HISTSIZE=1000
HOME=/home/wong
KDED_STARTED_BY_KDEINIT=1
KDE_APPLICATIONS_AS_SCOPE=1
KDE_FULL_SESSION=true
KDE_SESSION_UID=1000
KDE_SESSION_VERSION=5
KONSOLE_DBUS_SERVICE=:1.147
KONSOLE_DBUS_SESSION=/Sessions/1
KONSOLE_DBUS_WINDOW=/Windows/1
KONSOLE_VERSION=201200
LANG=en_US.UTF-8
LANGUAGE=''
LOGNAME=wong
LS_COLORS='rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:'
MAIL=/var/spool/mail/wong
MOTD_SHOWN=pam
MOZ_PLUGIN_PATH=/usr/lib/mozilla/plugins
OLDPWD=/home/wong
PAM_KWALLET5_LOGIN=/run/user/1000/kwallet5.socket
PATH=/usr/lib/ccache/bin/:/home/wong/go/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/opt/android-ndk:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
PROFILEHOME=''
PWD=/home/wong
QT_AUTO_SCREEN_SCALE_FACTOR=0
QT_IM_MODULE=fcitx
SAVEHIST=1000
SHELL=/usr/bin/zsh
SHELL_SESSION_ID=73674d586cb2440eb775f6e1829f212d
SHLVL=1
TERM=xterm-256color
USER=wong
WINDOWID=123731975
XAUTHORITY=/home/wong/.Xauthority
XCURSOR_SIZE=24
XCURSOR_THEME=breeze_cursors
XDG_CURRENT_DESKTOP=KDE
XDG_RUNTIME_DIR=/run/user/1000
XDG_SEAT=seat0
XDG_SEAT_PATH=/org/freedesktop/DisplayManager/Seat0
XDG_SESSION_CLASS=user
XDG_SESSION_DESKTOP=KDE
XDG_SESSION_ID=2
XDG_SESSION_PATH=/org/freedesktop/DisplayManager/Session1
XDG_SESSION_TYPE=x11
XDG_VTNR=7
XMODIFIERS='@im=fcitx'
XZ_OPT='--threads=0'
ZLSCOLORS='rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:'
```

### PR DESCRIPTION
**Pull request by iguessthislldo**
This is a backport of #13434, Fixes #13685.

I think builds using standalone toolchain are fine so I left them alone,
but `Configure` will fail if using the NDK directly because the
`platforms` and `sysroot` directories were removed.

If `sysroot` is missing, omit the `--sysroot` and `-gcc-toolchain`
arguments and use the triplet form clang command.

Also since `platforms` was being used for the default API level, use
`meta/platforms.json` instead if needed.

### CODE REVIEW

**Comment by wongsyrone** [comment]
I fixed it on my side using the patch. Works smoothly. Good job.

**Comment by tmm1** [comment]
Ping?

**Comment by SmallLars** [comment]
Would be nice to have this in 1.1.1j

**Comment by misery** [comment]
@levitte Could this be backported for NDK 22 (final), too?

**Comment by InfoHunter** [comment]
I consider this as a worthwhile backport, any thoughts on this @openssl/committers ?

File: Configurations/15-android.conf
Comment (vdukhovni) [inline Configurations/15-android.conf:32]: On line 27 the list of envvars checked is unchanged: `qw(ANDROID_NDK_HOME ANDROID_NDK)`, why is the error message changing from complaining about `ANDROID_NDK_HOME` to `ANDROID_NDK_ROOT`?

Comment (iguessthislldo) [reply Configurations/15-android.conf:32]: That's the variable name that is used on master and I didn't notice it was different. Thanks for pointing that out. I don't know why they're different, but I guess I'll put the old name back.

Comment (iguessthislldo) [reply Configurations/15-android.conf]: Done

**Comment by krushyu** [comment]
Issue remains in 1.1.1j, but this solution still work, thanks^^

**Comment by SmallLars** [comment]
Ping @ 1.1.1k

**Comment by changloong** [comment]
Maybe this patch should also add support for android-ndk-r23-beta2  ?

**Comment by changloong** [comment]
@iguessthislldo 

Maybe this patch should also add support for android-ndk-r23-beta2 ?

**Comment by iguessthislldo** [comment]
> Maybe this patch should also add support for android-ndk-r23-beta2 ?

Maybe, but I haven't looked at r23 at all yet. Is it having an issue that's different from the one with r22?

**Comment by changloong** [comment]
@iguessthislldo 
R23 size is 673MB compare to R22 1.1GB. It only provide llvm-11 based toolchains.

For `toolchains/llvm/prebuilt/linux-x86_64/bin` at R23:

```sh
ls *-ar
llvm-ar
```
at R22:
```sh
ls *-ar
aarch64-linux-android-ar  arm-linux-androideabi-ar  i686-linux-android-ar  llvm-ar  x86_64-linux-android-ar
```

I can not build with this patch for R23.

**Comment by iguessthislldo** [comment]
@changloong Okay, I am able to build this branch with r23-beta2 myself. So it's not using `llvm-ar` for you? As far as I can see it should be using that if `clang` from the NDK is the first thing on the `$PATH`. What `ar` did it try to use?

**Comment by SmallLars** [comment]
This is a backport of the NDK r22 support from OpenSSL 3.
So could we please focus on this and finally publish the NDK r22 support for 1.1.1?

NDK r23 is Beta and can be a different commit.

**Review by mattcaswell — APPROVED** [review]
I'm ok with this in principle. However, I think such a change would need OTC approval.

**Comment by changloong** [comment]
Hi @iguessthislldo , cloud be my env problem.  after I add this modify it work for me.

```patch
             } elsif ($user{CC} eq "clang") {
-                die "no NDK clang on \$PATH";
+                # die "no NDK clang on \$PATH";
             }
```

**Review by t-j-h — APPROVED** [review]

**Comment by mattcaswell** [comment]
OTC: This is ok for backport to 1.1.1

Lifting the OTC hold.

**Comment by mattcaswell** [comment]
Pushed. Thanks!

**Pull request closed without merging**
